external/fv3fit/fv3fit/pytorch/graph/mpg_unet.py

```diff
@@ -0,0 +1,331 @@
+                )
+                node_features, hidden_features = self.gru(
+                    node_features.unsqueeze(0), hidden_features
+                )
+                node_features = node_features.squeeze(0)
+            out_node_features[batch] = node_features.reshape(
+                in_size[0], in_size[1], in_size[2], node_features.size(1)
+            )  # reshape (tile * x * y, features) to (tile, x, y, features)
+        return out_node_features
+
+
+class Down(nn.Module):
+    """
+    A class for the down path of the U-net which
+    reduce the resolution by applying pooling layer
+    """
+
+    def __init__(self, pooling_size, pooling_stride):
+        super(Down, self).__init__()
+        self.pool = nn.AvgPool2d(kernel_size=pooling_size, stride=pooling_stride)
+
+    def forward(self, x):
+        batch_size, n_tiles, n_x, n_y, n_features = x.size()
+        x = x.permute(
+            0, 1, 4, 2, 3
+        )  # change dimensions to (batch_size, n_tiles, n_features, n_x, n_y)
+        x = x.reshape(x.size(0) * x.size(1), x.size(2), x.size(3), x.size(4))
+        x = self.pool(x)
+        x = x.reshape(batch_size, n_tiles, n_features, n_x // 2, n_y // 2,)
+
+        return x.permute(0, 1, 3, 4, 2)
+
+
+class Up(nn.Module):
+    """
+    A class for the processes on each level of up path of the U-Net
+    """
+
+    def __init__(self, pooling_size, pooling_stride, in_channels):
+        """
+        Args:
+            in_channels: size of input channels
+        """
+        super(Up, self).__init__()
+        self.up = nn.ConvTranspose2d(
+            in_channels,
+            in_channels // 2,
+            kernel_size=pooling_size,
+            stride=pooling_stride,
+        )
+
+    def forward(self, x1):
+        batch_size, n_tiles, n_x, n_y, n_features = x1.size()
+        x1 = x1.permute(
+            0, 1, 4, 2, 3
+        )  # change dimensions to (batch_size, n_tiles, n_features, n_x, n_y)
+        x1 = x1.reshape(
+            x1.size(0) * x1.size(1), x1.size(2), x1.size(3), x1.size(4)
+        )  # change the shape to (batch_size*n_tiles, n_features, n_x, n_y )
+        x1 = self.up(x1)
+        x1 = x1.reshape(batch_size, n_tiles, n_features // 2, n_x * 2, n_y * 2,)
+        x1 = x1.permute(
+            0, 1, 3, 4, 2
+        )  # change dimensions to (batch_size, n_tiles, n_x, n_y, n_features)
+        return x1
+
+
+class UNet(nn.Module):
+    """
+    A graph based U-net architucture
+    """
+
+    def __init__(
+        self, config, down_factory, up_factory, depth: int, in_channels: int, nx: int
+    ):
+        """
+        Args:
+            down_factory: double-convolution followed
+                by a pooling layer on the down-path side
+            up_factory: Upsampling followed by
+                double-convolution on the up-path side
+            depth: depth of the UNet
+            in_channels: number of input channels
```

Comment:
Please add a type hint for `config`.

The docstring doesn't match the arguments.